File: lfu/sol.java
package lfu;

import java.util.Arrays;
import java.util.Scanner;

/**
 * sol
 */
public class sol {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int capacity = Integer.parseInt(scanner.nextLine().trim());
        int[] weights = Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] uses = Arrays.stream(scanner.nextLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        System.out.println(solution(capacity, weights, uses));
        scanner.close();
    }

    public static int solution(int capacity, int[] weights, int[] uses) {
        // init
        int currentCapacity = 0;
        int[] frequencyMap = new int[weights.length];
        Arrays.fill(frequencyMap, 0);

        // algo
        for (int itemId : uses) {
            if (frequencyMap[itemId] == 0) {
                while (currentCapacity + weights[itemId] > capacity) {
                    // remove weight of item with least non-zero frequency
                    int minItemId = minNonZero(frequencyMap);
                    frequencyMap[minItemId] = 0;
                    currentCapacity = currentCapacity - weights[minItemId];
                }
                currentCapacity = currentCapacity + weights[itemId];
            }
            frequencyMap[itemId] = frequencyMap[itemId] + 1;
        }

        return currentCapacity;
    }

    public static int minNonZero(int[] array) {
        int min = 0;
        for (int i = 1; i < array.length; i++) {
            if (array[i] != 0 && array[i] < array[min])
                min = i;
        }

        return min;
    }
}

File: lfu/sol.js
const readline = require("readline").createInterface({ input: process.stdin });

const inputLines = [];
let index = 0;
const limit = 3; // 3 lines

readline.on("line", (line) => {
    inputLines.push(line);
    index++;
    if (index == limit) {
        readline.close();
        const input = {
            capacity: parseInt(inputLines[0]),
            weights: inputLines[1].split(" ").map((val) => parseInt(val)),
            uses: inputLines[2].split(" ").map((val) => parseInt(val)),
        };

        // solution
        let currentCapacity = 0;
        let frequencyMap = Array(6).fill(0); // init frequencies to 0
        for (const itemId of input.uses) {
            if (frequencyMap[itemId] === 0) {
                while (
                    currentCapacity + input.weights[itemId] >
                    input.capacity
                ) {
                    // remove weight of item with least non-zero frequency
                    const minItemId = frequencyMap.indexOf(
                        Math.min(...frequencyMap.filter((x) => x !== 0))
                    );
                    frequencyMap[minItemId] = 0;
                    currentCapacity -= input.weights[minItemId];
                }
                currentCapacity += input.weights[itemId];
            }
            frequencyMap[itemId]++;
        }

        console.log(currentCapacity);
    }
});
